pkg/controller/walle/controller.go

```diff
@@ -0,0 +1,435 @@
+	wr.logger.Log(
+		"msg", "starting run loop",
+		"threadiness", threadiness,
+		"v", 2,
+	)
+
+	wr.stopCh = stopCh
+	wr.informerWg = wg
+
+	defer wr.queue.ShutDown()
+	defer wg.Done()
+	wg.Add(1)
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(wr.runWorker, time.Second, stopCh)
+	}
+
+	ticker := time.NewTicker(KlusterRecheckInterval)
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				wr.requeueAllKlusters()
+			case <-stopCh:
+				ticker.Stop()
+				return
+			}
+		}
+	}()
+	<-stopCh
+}
+
+func (wr *WalleController) requeueAllKlusters() (err error) {
+	defer func() {
+		wr.logger.Log(
+			"msg", "requeued all",
+			"v", 1,
+			"err", err,
+		)
+	}()
+
+	klusters, err := wr.Factories.Kubernikus.Kubernikus().V1().Klusters().Lister().List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	for _, kluster := range klusters {
+		wr.requeueKluster(kluster)
+	}
+
+	return nil
+}
+
+func (wr *WalleController) requeueKluster(kluster *v1.Kluster) {
+	wr.logger.Log(
+		"msg", "queuing",
+		"kluster", kluster.Spec.Name,
+		"project", kluster.Account(),
+		"v", 2,
+	)
+
+	key, err := cache.MetaNamespaceKeyFunc(kluster)
+	if err == nil {
+		wr.queue.Add(key)
+	}
+}
+
+func (wr *WalleController) runWorker() {
+	for wr.processNextWorkItem() {
+	}
+}
+
+func (wr *WalleController) processNextWorkItem() bool {
+	key, quit := wr.queue.Get()
+	if quit {
+		return false
+	}
+	defer wr.queue.Done(key)
+
+	var err error
+	var kluster *v1.Kluster
+	var requeue bool
+
+	obj, exists, _ := wr.Factories.Kubernikus.Kubernikus().V1().Klusters().Informer().GetIndexer().GetByKey(key.(string))
+	if !exists {
+		err = errors.NewNotFound(schema.GroupResource{Resource: "kluster"}, "")
+	} else {
+		kluster = obj.(*v1.Kluster)
+	}
+
+	if err == nil {
+		// Invoke the method containing the business logic
+		requeue, err = wr.reconcile(kluster)
+	}
+
+	if err == nil {
+		// Forget about the #AddRateLimited history of the key on every successful synchronization.
+		// This ensures that future processing of updates for this key is not delayed because of
+		// an outdated error history.
+
+		if requeue == false {
+			wr.queue.Forget(key)
+		} else {
+			// Requeue requested
+			wr.queue.AddAfter(key, BaseDelay)
+		}
+
+		return true
+	}
+
+	if wr.queue.NumRequeues(key) < 5 {
+		// Re-enqueue the key rate limited. Based on the rate limiter on the
+		// queue and the re-enqueue history, the key will be processed later again.
+		wr.queue.AddRateLimited(key)
+		return true
+	}
+
+	// Retries exceeded. Forgetting for this reconciliation loop
+	wr.queue.Forget(key)
+	return true
+}
+
+func (wr *WalleController) reconcile(kluster *v1.Kluster) (bool, error) {
+
+	if err := wr.createNodeInformerForKluster(kluster); err != nil {
+		return true, err
+	}
+
+	if err := wr.cleanUpInformers(); err != nil {
+		return true, err
+	}
+
+	if err := wr.watchNodesForKluster(kluster); err != nil {
```

Comment:
I'm not sure this is the right place for this method. IMO this should be done in `createNodeInformerForKluster` when the `NodeInformer` struct is actually created.

Reply:
I wanted the cleanup to happen before we're watching. 
Also makes sense to change as you suggested.